docs/features/task-types.md

```diff
@@ -0,0 +1,90 @@
+Async Task Types in C#
+======================
+Extend `async` to support any _task type_ that matches a specific pattern, not just the well known types
+`Sytem.Threading.Tasks.Task` and `Sytem.Threading.Tasks.Task<T>`.
+
+## Task Type
+A _task type_ is a `class` or `struct` with an associated _builder type_ identified
+with `System.Runtime.CompilerServices.AsyncMethodBuilderAttribute`.
+
+A _task type_ can be non-generic, for async methods that do not return a value, or generic, for methods that return a value.
+Generic _task types_ have a `Result` property.
+
+To support `await`, the _task type_ must have a corresponding, accessible `GetAwaiter()` method.
+```
+[AsyncMethodBuilder(typeof(MyTaskMethodBuilder<>))]
+class MyTask<T>
+{
+    public T Result { get; }
+    public Awaiter<T> GetAwaiter();
+}
+```
+## Awaiter Type
+The _awaiter type_ is a `class` or `struct` that implements `System.Runtime.CompilerServices.ICriticalNotifyCompletion` or `System.Runtime.CompilerServices.INotifyCompletion`. 
+```
+class Awaiter<T> : INotifyCompletion
+{
+    public void OnCompletion(Action completion);    
+}
+```
+## Builder Type
+The _builder type_ is a `class` or `struct` that corresponds to a specific _task type_, with the following `public` methods.
+For non-generic _builder types_, `SetResult()` has no parameters.
+```
+class MyTaskMethodBuilder<T>
+{
+    public static MyTaskMethodBuilder<T> Create();
+
+    public void Start<TStateMachine>(ref TStateMachine stateMachine)
+        where TStateMachine : IAsyncStateMachine;
+
+    public void SetStateMachine(IAsyncStateMachine stateMachine);
+    public void SetException(Exception exception);
+    public void SetResult(T result);
+
+    public void AwaitOnCompleted<TAwaiter, TStateMachine>(
+        ref TAwaiter awaiter, ref TStateMachine stateMachine)
+        where TAwaiter : INotifyCompletion
+        where TStateMachine : IAsyncStateMachine;
+    public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(
+        ref TAwaiter awaiter, ref TStateMachine stateMachine)
+        where TAwaiter : ICriticalNotifyCompletion
+        where TStateMachine : IAsyncStateMachine;
+
+    public MyTask<T> Task { get; }
+}
+```
+## Execution
```

Comment:
I don't know how to phrase this, but it may be good to mention this is the same design used by `Task` and `Task<T>`, which use `System.Runtime.CompilerServices.AsyncTaskMethodBuilder` and `System.Runtime.CompilerServices.AsyncTaskMethodBuilder<TResult>` respectively. The main difference with those known types, is that the compiler already knows which async method builder to use.